File: ReactAndroid/src/main/java/com/facebook/react/bridge/Inspector.java
// Copyright (c) Facebook, Inc. and its affiliates.

// This source code is licensed under the MIT license found in the
// LICENSE file in the root directory of this source tree.

package com.facebook.react.bridge;

import com.facebook.common.logging.FLog;
import com.facebook.jni.HybridData;
import com.facebook.proguard.annotations.DoNotStrip;
import com.facebook.react.common.ReactConstants;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

@DoNotStrip
public class Inspector {
  static {
    ReactBridge.staticInit();
  }

  private final HybridData mHybridData;

  public static List<Page> getPages() {
    try {
      return Arrays.asList(instance().getPagesNative());
    } catch (UnsatisfiedLinkError e) {
      FLog.e(ReactConstants.TAG, "Inspector doesn't work in open source yet", e);
      return Collections.emptyList();
    }
  }

  public static LocalConnection connect(int pageId, RemoteConnection remote) {
    try {
      return instance().connectNative(pageId, remote);
    } catch (UnsatisfiedLinkError e) {
      FLog.e(ReactConstants.TAG, "Inspector doesn't work in open source yet", e);
      throw new RuntimeException(e);
    }
  }

  private static native Inspector instance();

  private native Page[] getPagesNative();

  private native LocalConnection connectNative(int pageId, RemoteConnection remote);

  private Inspector(HybridData hybridData) {
    mHybridData = hybridData;
  }

  @DoNotStrip
  public static class Page {
    private final int mId;
    private final String mTitle;
    private final String mVM;

    public int getId() {
      return mId;
    }

    public String getTitle() {
      return mTitle;
    }

    public String getVM() {
      return mVM;
    }

    @Override
    public String toString() {
      return "Page{" + "mId=" + mId + ", mTitle='" + mTitle + '\'' + '}';
    }

    @DoNotStrip
    private Page(int id, String title, String vm) {
      mId = id;
      mTitle = title;
      mVM = vm;
    }
  }

  @DoNotStrip
  public interface RemoteConnection {
    @DoNotStrip
    void onMessage(String message);

    @DoNotStrip
    void onDisconnect();
  }

  @DoNotStrip
  public static class LocalConnection {
    private final HybridData mHybridData;

    public native void sendMessage(String message);

    public native void disconnect();

    private LocalConnection(HybridData hybridData) {
      mHybridData = hybridData;
    }
  }
}

File: packages/react-native/ReactAndroid/src/main/java/com/facebook/react/bridge/Inspector.kt
/*
 * Copyright (c) Meta Platforms, Inc. and affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

package com.facebook.react.bridge

import com.facebook.common.logging.FLog
import com.facebook.jni.HybridData
import com.facebook.proguard.annotations.DoNotStrip
import com.facebook.react.common.ReactConstants

@DoNotStrip
public class Inspector
private constructor(@Suppress("NoHungarianNotation") private val mHybridData: HybridData) {

  private external fun getPagesNative(): Array<Page>

  private external fun connectNative(pageId: Int, remote: RemoteConnection): LocalConnection?

  @DoNotStrip
  public class Page
  private constructor(private val id: Int, private val title: String, private val vm: String) {
    public fun getId(): Int = id

    public fun getTitle(): String = title

    public fun getVM(): String = vm

    override fun toString(): String = "Page{id=$id, title='$title'}"
  }

  @DoNotStrip
  public interface RemoteConnection {
    @DoNotStrip public fun onMessage(message: String)

    @DoNotStrip public fun onDisconnect()
  }

  @DoNotStrip
  public class LocalConnection
  private constructor(@Suppress("NoHungarianNotation") private val mHybridData: HybridData) {
    public external fun sendMessage(message: String)

    public external fun disconnect()
  }

  public companion object {
    init {
      BridgeSoLoader.staticInit()
    }

    @JvmStatic
    public fun getPages(): List<Page> {
      return try {
        instance().getPagesNative().toList()
      } catch (e: UnsatisfiedLinkError) {
        FLog.e(ReactConstants.TAG, "Inspector doesn't work in open source yet", e)
        emptyList()
      }
    }

    @JvmStatic
    public fun connect(pageId: Int, remote: RemoteConnection): LocalConnection {
      return try {
        instance().connectNative(pageId, remote)
            ?: throw IllegalStateException("Can't open failed connection")
      } catch (e: UnsatisfiedLinkError) {
        FLog.e(ReactConstants.TAG, "Inspector doesn't work in open source yet", e)
        throw RuntimeException(e)
      }
    }

    @JvmStatic private external fun instance(): Inspector
  }
}
